engine/src/main/java/com/sandboni/core/engine/sta/executor/AbstractParallelExecutor.java

```diff
@@ -0,0 +1,52 @@
+package com.sandboni.core.engine.sta.executor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+
+import static com.sandboni.core.engine.utils.TimeUtils.elapsedTime;
+import static java.util.stream.Collectors.toList;
+
+public abstract class AbstractParallelExecutor<T, R> implements ParallelExecutor<Collection<T>, Collection<R>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractParallelExecutor.class);
+
+    @Override
+    public Collection<R> execute(Collection<T> input) {
+        long start = System.nanoTime();
+
+        List<CompletableFuture<R>> futures = input.stream()
+                .map(getLoggableMappingFunction())
+                .collect(toList());
+
+        List<R> result = futures.stream()
+                .map(CompletableFuture::join)
+                .collect(toList());
+
+        logger.info("[{}] {} executed in {} milliseconds", Thread.currentThread().getName(), this.getClass().getSimpleName(), elapsedTime(start));
+
+        return result;
+    }
+
+    abstract Function<T, R> getMappingFunction();
+
+    protected Function<T, CompletableFuture<R>> getLoggableMappingFunction() {
+        return inputItem -> {
+            logger.debug("Accepting input {}", inputItem);
+            return CompletableFuture.supplyAsync(() -> {
+                logger.debug("Thread {} Processing inputItem {}", Thread.currentThread().getName(), inputItem);
+
+                return getMappingFunction().apply(inputItem);
+            }, getExecutorPool());
```

Comment:
What is the benefit here? managing the thread pool ourselves as opposed to paralleling the stream and letting java decide? how did you measure the difference in performance?

Reply:
Using parallelStream is not the best approach when there's I/O involved because can potentially slow down the whole JVM (as there's only one shared ForkJoinPool). Instead, another thread pool can be created where we can decide the number of threads. 
I used profiling to measure the performance.